src/app/core/utils/math/Graph.ts

```diff
@@ -97,4 +97,76 @@ export class Graph<V, E> {
         return nodes;
     }
 
+    /**
+     * Used to get the max "depth" of each node where the source nodes have a depth of 0
+     *  and each additional layer adds +1 to the depth. If a node has parents at multiple
+     *  depths such as 1 and 3, then it inherits from the "deeper" one, so the node would
+     *  have a depth of 4. This should only be called if the isConnected() is true and the
+     *  graph is acyclic.
+     *
+     * @return a map where each key is each node and the value is the max depth of that node
+     */
+    public getMaxNodeDepths(): Map<V, number> {
+        const sources = this.getSources();
+        const ret = new Map<V, number>();
+
+        let currentLayer = this.getSources();
+        let nextLayer: V[] = [];
+
+        for (const node of currentLayer) {
+            ret.set(node, 0);
+        }
+
+        while (currentLayer.length != 0) {
+            for (const node of currentLayer) {
+                const nextDepth = ret.get(node) + 1;
+                for (const next of this.list.get(node))  {
+                    ret.set(next.getTarget(), nextDepth);
+                    nextLayer.push(next.getTarget());
+                }
+            }
+            currentLayer = nextLayer;
+            nextLayer = [];
+        }
+
+        return ret;
+    }
+
+    /**
+     * Used to get the min "depth" of each node where the source nodes have a depth of 0
+     *  and each additional layer adds +1 to the depth. If a node has parents at multiple
+     *  depths such as 1 and 3, then it inherits from the "shallower" one, so the node would
+     *  have a depth of 2. This should only be called if the isConnected() is true and the
+     *  graph is acyclic.
+     *
+     * @return a map where each key is each node and the value is the max depth of that node
+     */
+    public getMinNodeDepths(): Map<V, number> {
+        const sources = this.getSources();
+        const ret = new Map<V, number>();
+
+        let currentLayer = this.getSources();
+        let nextLayer: V[] = [];
+
+        for (const node of currentLayer) {
+            ret.set(node, 0);
+        }
+
+        while (currentLayer.length != 0) {
+            for (const node of currentLayer) {
+                const nextDepth = ret.get(node) + 1;
+                for (const next of this.list.get(node))  {
+                    if (!ret.has(next.getTarget())) {
+                        ret.set(next.getTarget(), nextDepth);
+                        nextLayer.push(next.getTarget());
+                    }
+                }
+            }
+            currentLayer = nextLayer;
+            nextLayer = [];
+        }
+
+        return ret;
+    }
+
```

Comment:
Condensed down so that they both call a common privated function. Also, I think the bfs function currently in the Graph class is actually a depth first search, but since it only indicates to the caller whether a node is found or not I don't think it would effectively make a difference. Either way, I decided not to modify it since it also assumes the graph is undirected, which for my purpose it isn't (but it is when isConnected() is called).